### PR TITLE
feat(ha_nexus): adding support to handle switchover request asynchronously

### PR DESCRIPTION
**Pull request by abhilashshetty04**
This PR adds support to create worker tasks to process switchover requests parallely.

Signed-off-by: Abhilash Shetty <abhilash.shetty@datacore.com>

### CODE REVIEW

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 1]
```diff
@@ -243,48 +247,55 @@ impl SwitchOverEngine {
...
-        });
+        }
+    }
+
+    /// Mover instantiates worker task to asynchronously process Switchover request.
+    pub fn mover(&self, recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>) {
+        for i in 0 .. WORKER_NUM {
+            info!("Starting worker-{} of Switchover Engine", i.to_string());
+            let cloned_self = self.clone();
+            let cloned_channel = Arc::clone(&recv);
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: Clone takes ref self so you can just call clone:
```suggestion
            let cloned_channel = recv.clone();
```

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 2]
```diff
@@ -243,48 +247,55 @@ impl SwitchOverEngine {
...
-                        match result {
-                            Ok(_) => {
-                                // reset retry delay back to the start.
-                                retry_delay = retry_delay_min;
-                            }
-                            Err(_) => {
-                                retry_delay = (retry_delay * 3).min(retry_delay_max);
-                                tokio::time::sleep(retry_delay).await;
-                            }
+        loop {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs:266]: You don't need to clone etcd and nodes now?

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 3]
```diff
@@ -243,48 +247,55 @@ impl SwitchOverEngine {
...
+        for i in 0 .. WORKER_NUM {
+            info!("Starting worker-{} of Switchover Engine", i.to_string());
+            let cloned_self = self.clone();
+            let cloned_channel = Arc::clone(&recv);
+            tokio::spawn(async move { cloned_self.process_request(cloned_channel).await });
+        }
+    }
+
+    /// Switchover request to be handled synchronously in each worker task.
+    async fn process_request(&self, recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>) {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]: think we can just consume self here:
```suggestion
    async fn process_request(self, recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>) {
```

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 4]
```diff
@@ -232,6 +233,9 @@ pub struct SwitchOverEngine {
     channel: UnboundedSender<SwitchOverRequest>,
 }
 
+/// Num of Switchover instances to be spawned.
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs]:
```suggestion
/// Number of Switchover instances to be spawned.
```

**Review by tiagolobocastro — APPROVED** [review]
A possible problem with this approach is that the work per worker is still blocking, so if one volume gets stuck for whatever reason, we block that whole Switchover instance.
Let's merge this as it's already better but I think we still need a solution for that..

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 3 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Incorporated

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 2 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs:266]: Incorporated

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 1 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Incorporated

File: control-plane/agents/src/bin/ha/cluster/switchover.rs (hunk 4 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs]: Incorporated

File: control-plane/agents/src/bin/ha/cluster/switchover.rs [hunk 5]
```diff
@@ -243,48 +247,53 @@ impl SwitchOverEngine {
...
+            let cloned_self = self.clone();
+            let cloned_channel = recv.clone();
+            tokio::spawn(async move { cloned_self.process_request(cloned_channel).await });
+        }
+    }
+
+    /// Switchover request to be handled synchronously in each worker task.
+    async fn process_request(self, recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>) {
+        loop {
+            if let Some(mut q) = recv.lock().await.recv().await {
```
Comment (abhilashshetty04) [inline control-plane/agents/src/bin/ha/cluster/switchover.rs:267]: @tiagolobocastro , Will this create lock contention?

Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs:267]: I guess not.

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs:267]: It's just taking switchover requests out of the queue, it shouldn't be a problem.

Comment (abhilashshetty04) [reply control-plane/agents/src/bin/ha/cluster/switchover.rs:267]: Thanks

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**